src/bulkUtils.ts

```diff
@@ -174,3 +182,125 @@ export const remainingTime =
   (now: number) =>
   (endWaitTime?: number): number =>
     Math.max((endWaitTime ?? now) - now, 0);
+
+export enum ColumnDelimiter {
+  BACKQUOTE = '`',
+  CARET = '^',
+  COMMA = ',',
+  PIPE = '|',
+  SEMICOLON = ';',
+  TAB = '	',
+}
+
+export type ColumnDelimiterKeys = keyof typeof ColumnDelimiter;
+
+async function bulkRequest(conn: Connection, url: string): Promise<{ body: string; headers: HttpResponse['headers'] }> {
+  const httpApi = new HttpApi(conn, {
+    responseType: 'text/plain', // this ensures jsforce doesn't try parsing the body
+  });
+
+  let headers: HttpResponse['headers'] | undefined;
+
+  httpApi.on('response', (response: HttpResponse) => {
+    headers = response.headers;
+  });
+
+  const body = await httpApi.request<string>({
+    url: conn.normalizeUrl(url),
+    method: 'GET',
+  });
+
+  if (!headers) throw new Error('failed to get HTTP headers for bulk query');
+
+  return {
+    body,
+    headers,
+  };
+}
+
+export async function exportRecords(
+  conn: Connection,
+  queryJob: QueryJobV2<Schema>,
+  outputInfo: {
+    filePath: string;
+    format: 'csv' | 'json';
+    columnDelimiter: ColumnDelimiterKeys;
+  }
+): Promise<QueryJobInfoV2> {
+  let jobInfo: QueryJobInfoV2 | undefined;
+
+  queryJob.on('jobComplete', (completedJob: QueryJobInfoV2) => {
+    jobInfo = completedJob;
+  });
+
+  await queryJob.poll();
+
+  if (jobInfo === undefined) {
+    throw new Error('could not get job info after polling');
+  }
+
+  let locator: string | undefined;
+
+  let recordsWritten = 0;
+
+  while (locator !== 'null') {
+    // we can't parallelize this because we:
+    // 1. need to get 1 batch to know the locator for the next one
+    // 2. merge all batches into one csv or json file
+    //
+    // eslint-disable-next-line no-await-in-loop
+    const res = await bulkRequest(
+      conn,
+      locator ? `/jobs/query/${jobInfo.id}/results?locator=${locator}` : `/jobs/query/${jobInfo.id}/results`
+    );
+
+    if (outputInfo.format === 'json') {
+      const jsonWritable = fs.createWriteStream(outputInfo.filePath, {
+        flags: 'a', // append mode. TODO: maybe make this append + fail if path exists?
+      });
+
+      const totalRecords = jobInfo.numberRecordsProcessed;
+
+      if (!locator) {
+        // first write, start JSON array
+        jsonWritable.write(`[${EOL}`);
+      }
+
+      // eslint-disable-next-line no-await-in-loop
+      await pipeline(
+        Readable.from(res.body),
+        new csvParse({ columns: true, delimiter: ColumnDelimiter[outputInfo.columnDelimiter] }),
```

Comment:
same options as jsforce (`columns: true`):
https://github.com/jsforce/jsforce/blob/main/src/csv.ts

`sf data export bulk` allows to specify the delimiter so we pass it here for csv parsing.